pysmt/test/test_unsat_cores.py

```diff
@@ -0,0 +1,150 @@
+from pysmt.exceptions import SolverStatusError
+from pysmt.test.examples import get_example_formulae
+
+class TestUnsatCores(TestCase):
+
+    @unittest.skipIf(len(get_env().factory.all_unsat_core_solvers()) == 0,
+                     "No Solver supporting Unsat Cores is available.")
+    def test_basic(self):
+        x = Symbol("x")
+        with UnsatCoreSolver(logic=QF_BOOL) as s:
+            s.add_assertion(x)
+            s.add_assertion(Not(x))
+            r = s.solve()
+            self.assertFalse(r)
+
+            core = s.get_unsat_core()
+            self.assertEqual(len(core), 2)
+            self.assertIn(x, core)
+            self.assertIn(Not(x), core)
+
+            named_core = s.get_named_unsat_core()
+            self.assertEqual(len(core), 2)
+            self.assertIn(x, named_core.values())
+            self.assertIn(Not(x), named_core.values())
+
+
+    @unittest.skipIf(len(get_env().factory.all_unsat_core_solvers()) == 0,
+                     "No Solver supporting Unsat Cores is available.")
+    def test_basic_named(self):
+        x = Symbol("x")
+        with UnsatCoreSolver(logic=QF_BOOL, unsat_cores_mode="named") as s:
+            s.add_assertion(x, named="a1")
+            s.add_assertion(Not(x), named="a2")
+            r = s.solve()
+            self.assertFalse(r)
+
+            core = s.get_unsat_core()
+            self.assertEqual(len(core), 2)
+            self.assertIn(x, core)
+            self.assertIn(Not(x), core)
+
+            named_core = s.get_named_unsat_core()
+            self.assertEqual(len(named_core), 2)
+            self.assertIn("a1", named_core)
+            self.assertIn("a2", named_core)
+            self.assertEqual(named_core["a1"], x)
+            self.assertEqual(named_core["a2"], Not(x))
+
+
+    @unittest.skipIf(len(get_env().factory.all_unsat_core_solvers()) == 0,
+                     "No Solver supporting Unsat Cores is available.")
+    def test_modify_state(self):
+        x = Symbol("x")
+        with UnsatCoreSolver(logic=QF_BOOL) as s:
+            s.add_assertion(x)
+            s.push()
+            s.add_assertion(Not(x))
+            r = s.solve()
+            self.assertFalse(r)
+            s.pop()
+            with self.assertRaises(SolverStatusError):
+                s.get_unsat_core()
+
+
+    @unittest.skipIf(len(get_env().factory.all_unsat_core_solvers()) == 0,
+                     "No Solver supporting Unsat Cores is available.")
+    def test_modify_state_assert(self):
+        x = Symbol("x")
+        with UnsatCoreSolver(logic=QF_BOOL) as s:
+            s.add_assertion(x)
+            s.add_assertion(Not(x))
+            r = s.solve()
+            self.assertFalse(r)
+            s.add_assertion(Symbol("y"))
+            with self.assertRaises(SolverStatusError):
+                s.get_unsat_core()
+
+
+    @skipIfSolverNotAvailable("msat")
+    def test_examples_msat(self):
+        for (f, _, satisfiability, logic) in get_example_formulae():
```

Comment:
Factor this out, in a method that takes the target solver. This will allow to unify test_examples_msat and test_examples_z3.